### PR TITLE
msgidparse: make whitespace separating message ids optional

### LINKED ISSUE #72: msgidparse: assumes References are separated by whitespace, but it is optional
The current implementation splits ids within the `References` header on whitespace: https://github.com/staktrace/mailparse/blob/92ada51009052638d3e0f188219b89008ebe5208/src/msgidparse.rs#L56

However, [RFC5322 states](https://tools.ietf.org/html/rfc5322#section-3.6.4) (emphasis mine):
> The "References:" and "In-Reply-To:" fields each contain one or more unique message identifiers, **optionally** separated by CFWS [whitespace].

We're hitting this edge case in some emails pulled from an Exchange server.

I'd like to work on a solution to this initially, if that's okay.

---

Edit, for clarity, a failing test in `0.12.1` looks like:
```rust
        assert_eq!(
            msgidparse("<msg_one@foo.com><msg_two@bar.com>").unwrap(),
            MessageIdList(vec![
                "msg_one@foo.com".to_string(),
                "msg_two@bar.com".to_string(),
            ])
        );
```

### PR DESCRIPTION
**Pull request by tommilligan**
Closes #72 

This PR:

- adds a benchmark using [criterion](https://github.com/bheisler/criterion.rs) for the current `msgidparse` function
- covers the existing `msgidparse` function with unit tests
- updates the implementation of `msgidparse` to make separator whitespace optional, as per  [RFC5322](https://tools.ietf.org/html/rfc5322#section-3.6.4) 

## Benchmark results

**No significant decrease** in speed was seen for short or long inputs to `msgidparse`.

### master

```log
msgidparse single id    time:   [74.461 ns 74.700 ns 74.958 ns]
msgidparse many ids     time:   [8.0999 us 8.1436 us 8.1853 us]
```

### This PR

```log
msgidparse single id    time:   [75.269 ns 75.839 ns 76.397 ns]
                        change: [+0.1514% +0.8757% +1.6172%] (p = 0.02 < 0.05)
                        Change within noise threshold.
msgidparse many ids     time:   [8.1968 us 8.2692 us 8.3627 us]
                        change: [-0.1450% +1.0312% +2.5808%] (p = 0.15 > 0.05)
                        No change in performance detected.
```

### CODE REVIEW

File: src/msgidparse.rs [hunk 1]
```diff
@@ -53,14 +53,86 @@ impl fmt::Display for MessageIdList {
 /// ```
 pub fn msgidparse(ids: &str) -> Result<MessageIdList, MailParseError> {
     let mut msgids = Vec::new();
-    for id in ids.split_whitespace() {
-        if !id.starts_with('<') {
+
+    // The remaining section of the header, not yet chomped
+    let mut remaining = ids.clone();
+    // Remove any leading whitespace, after this we just trim trailing whitespace
+    remaining = remaining.trim_start();
```
Comment (staktrace) [inline src/msgidparse.rs:60]: The `clone()` here seems unnecessary. Can you just do `let mut remaining = ids.trim_start();`?

File: Cargo.toml [hunk 2]
```diff
@@ -21,3 +21,10 @@ maintenance = { status = "passively-maintained" }
 base64 = "0.12"
 quoted_printable = "0.4.2"
 charset = "0.1.1"
+
+[dev-dependencies]
+criterion = "0.3"
```
Comment (staktrace) [inline Cargo.toml:26]: While I appreciate the performance testing, I don't want to land anything that adds extra dependencies or dev-dependencies to mailparse. I feel pretty strongly about keeping the dependency graph/trust chain small, and anything not absolutely essential shouldn't be here. So I'll happily merge the two commits that actually fix the issue (with my other review comment addressed) but I'd like to skip this benchmarking stuff.

Comment (tommilligan) [reply Cargo.toml:26]: Sure, no problem. I just figured this is a performance-critical crate, so didn't want to cause a regression in the (more likely) existing cases.

File: src/msgidparse.rs (hunk 1 above)
Comment (tommilligan) [reply src/msgidparse.rs:60]: Good catch, thanks

**Comment by tommilligan** [comment]
@staktrace updated with your comments addressed.
